python/user/bot/user_bot.py

```diff
@@ -26,29 +45,82 @@ def main():
     logging.info(f'starting a the user_bot for party {party}')
     client = network.aio_party(party)
 
+    # default to 14 days as defined in DAML
+    archive_state = ArchiveState(message_heap=[], archive_after=1209600)
+
+    def expired(after: int, posted_at: int) -> bool:
+        return (datetime.datetime.fromtimestamp(posted_at) + datetime.timedelta(
+            seconds=after)) < datetime.datetime.now()
+
     @client.ledger_ready()
-    def bot_ready(_):
-        logging.info(f'user_bot for party {party} is ready')
+    async def bot_ready(_):
+        existing_messages = client.find_active(Chat.Message, {'sender': client.party})
+        archive_state.message_heap = [Message(int(cdata['postedAt']), cid)
+                                      for cid, cdata in existing_messages.items()]
+
+        message_heap = archive_state.message_heap
+        heapq.heapify(message_heap)
+        logging.info(f"Message cache loaded.")
+
+        (_, settings_cdata) = await client.find_one(Chat.UserSettings, {'user': client.party})
+        archive_state.archive_after = (settings_cdata['archiveMessagesAfter'])
+
+        # initial scan
+        while len(message_heap) > 0:
+            top = message_heap[0]
+            if not expired(archive_state.archive_after, top.post_at):
+                logging.info("Auto-archiving caught up.")
+                break
+            logging.info(f"Auto-archiving {top.cid} posted at {top.post_at}")
+            await client.submit(exercise(top.cid, 'Archive'))
+            heapq.heappop(message_heap)
+
+        logging.info(f'started auto-archiving bot for party {party}')
+        while True:
+            try:
+                while len(message_heap) > 0 and expired(archive_state.archive_after,
+                                                        message_heap[0].post_at):
+                    top = message_heap[0]
+                    logging.info(f'archiving {Chat.Message}:{top.cid}'
+                                 f' expired after {archive_state.archive_after}s')
+                    await client.submit(exercise(top.cid, 'Archive'))
```

Comment:
If this `cid` has been archived by some other process, then this will fail _and leave that cid on the heap_, ensuring the next iteration through the loop will also fail on the same CID, no?